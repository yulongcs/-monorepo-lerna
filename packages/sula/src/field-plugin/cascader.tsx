import React from 'react';
import { Cascader as ACascader } from 'antd';
import { CascaderProps as ACascaderProps } from 'antd/lib/cascader';

export type CascaderSourceItem = {
  text: any;
  value: any;
};

export interface CascaderSource {
  text: any;
  value: any;
  children: CascaderSourceItem[];
}

export interface CascaderProps extends ACascaderProps {
  source: CascaderSource[];
}

export default class Cascader extends React.Component<CascaderProps> {
  static defaultProps = {
    fieldNames: {
      label: 'text',
    },
  };

  render() {
    const { source = [], ...restProps } = this.props;
    return <ACascader options={source} {...restProps} />;
  }
}
