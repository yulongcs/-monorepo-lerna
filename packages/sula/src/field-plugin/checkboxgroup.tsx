import React from 'react';
import { Checkbox } from 'antd';

const CheckboxGroup = Checkbox.Group;

export default class SulaCheckboxGroup extends React.Component {
  render() {
    const { source = [], ...restProps } = this.props;
    return (
      <CheckboxGroup
        {...restProps}
        options={source.map((item) => {
          return {
            ...item,
            label: item.text,
          };
        })}
      />
    );
  }
}
