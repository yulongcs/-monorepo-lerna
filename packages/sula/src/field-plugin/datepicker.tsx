import React from 'react';
import moment from 'moment';
import { DatePicker } from 'antd';
import { DatePickerProps } from 'antd/lib/date-picker';


export type SulaDatePickerProps = DatePickerProps & {
  valueFormat?: 'utc' | boolean;
}

export default class SulaTimePicker extends React.Component<SulaDatePickerProps> {
  static defaultProps = {
    format: 'YYYY-MM-DD',
  }
  onFormatChange = (time, timeString) => {
    const {
      valueFormat,
      onChange,
    } = this.props;
    let finalTime;

    if(valueFormat === 'utc') {
      finalTime = time ? time.valueOf() : null;
    } else {
      finalTime = timeString;
    }

    if(onChange) {
      onChange(finalTime, timeString);
    }
    
  }
  render() {
    const {
      valueFormat, value, ...restProps
    } = this.props;
    let finalValue = value;
    if(value) {
      if(valueFormat === true) { // dateString
        finalValue = moment(value, restProps.format);
      } else {
        finalValue = moment(value);
      }
    }
    return (
      <DatePicker 
      {...restProps}
      {...(valueFormat ? { onChange: this.onFormatChange } : {})}
      value={finalValue}
      />
    )
  }
}