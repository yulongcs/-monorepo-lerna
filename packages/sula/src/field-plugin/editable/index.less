@import (reference) '~antd/lib/style/themes/default';

.sula-editable {
  & .ant-table-tbody > tr {
    vertical-align: baseline;

    & > td {
      padding-bottom: 0;
      padding-top: 24px;
    }
  }

  &-view {
    & .ant-table-tbody > tr {
      vertical-align: baseline;
  
      & > td {
        padding: 16px;
      }
    }
  }
}


