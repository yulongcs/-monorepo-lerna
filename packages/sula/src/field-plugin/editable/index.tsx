import React from 'react';
import { Table, Space, Button } from 'antd';
import { PlusOutlined } from '@ant-design/icons';
import cx from 'classnames';
import { Field, FieldProps } from '../../form';
import { Dependencies, DependencyType } from '../../types/dependency';
import { toArray } from '../../_util/common';
import { FieldNameList, FieldNamePath } from '../../types/form';
import LocaleReceiver from '../../localereceiver';
import './index.less';

interface FieldData {
  name: number;
  key: number;
  fieldKey: number;
  [key: string]: any;
}

export interface EditableProps {
  fields: FieldProps[];
  name: FieldNamePath;
}

export default class Editable extends React.Component<EditableProps> {
  static defaultProps = {
    fields: [],
  };

  renderField = (field: FieldData, fieldConfig: FieldProps, isViewMode: boolean) => {
    const { name, key, fieldKey, ...restField } = field;
    const { dependency = {} as Dependencies, label, ...restFieldConfig } = fieldConfig;
    const transformedDep = Object.keys(dependency).reduce(
      (memo: Dependencies, depType: DependencyType) => {
        const dep = dependency[depType];
        const relates = dep!.relates.map((r) => {
          return [fieldKey, ...toArray(r)];
        });
        memo[depType] = {
          ...dep,
          relates,
        };
        return memo;
      },
      {},
    ) as Dependencies;

    const fieldName = [name, ...toArray(fieldConfig.name)] as FieldNameList;
    const fieldNameKey = [fieldKey, ...toArray(fieldConfig.name)];

    return (
      <Field
        {...restField}
        {...restFieldConfig}
        noStyle={isViewMode}
        key={key}
        name={fieldName}
        fieldKey={fieldNameKey}
        dependency={transformedDep}
      />
    );
  };

  render() {
    const { list, ctx } = this.props;
    const [fields, { add, remove }] = list;
    const { mode } = ctx;
    const isViewMode = mode === 'view';
    return (
      <LocaleReceiver>
        {(locale) => {
          return (
            <div>
              <Table<FieldData>
                className={cx('sula-editable', {
                  [`sula-editable-view`]: isViewMode,
                })}
                pagination={false}
                dataSource={fields}
                rowKey="fieldKey"
                columns={[
                  ...this.props.fields.map((fieldConfig) => {
                    return {
                      title: fieldConfig.label,
                      key: toArray(fieldConfig.name).join('_'),
                      width: fieldConfig.width,
                      render: (_: any, field: FieldData) => {
                        return this.renderField(field, fieldConfig, isViewMode);
                      },
                    };
                  }),
                  ...(isViewMode
                    ? []
                    : [
                        {
                          title: '',
                          width: 160,
                          key: 'operation',
                          render: (_, record) => {
                            return (
                              <Space>
                                <a
                                  onClick={() => {
                                    add({}, record.name + 1);
                                  }}
                                >
                                  {locale.addText}
                                </a>
                                <a
                                  onClick={() => {
                                    const values = ctx.form.getFieldValue([
                                      ...toArray(this.props.name),
                                      record.name,
                                    ]);
                                    add(values, record.name + 1);
                                  }}
                                >
                                  {locale.copyText}
                                </a>
                                <a
                                  onClick={() => {
                                    remove(record.name);
                                  }}
                                >
                                  {locale.deleteText}
                                </a>
                              </Space>
                            );
                          },
                        },
                      ]),
                ]}
              />
              {isViewMode ? null : (
                <Button
                  icon={<PlusOutlined />}
                  block
                  type="dashed"
                  style={{ marginTop: 16 }}
                  onClick={() => {
                    add();
                  }}
                >
                  <span>{locale.addText}</span>
                </Button>
              )}
            </div>
          );
        }}
      </LocaleReceiver>
    );
  }
}
