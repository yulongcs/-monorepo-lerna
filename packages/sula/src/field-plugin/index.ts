import Select from './select';
import Upload from './upload';
import Cascader from './cascader';
import TreeSelect from './treeselect';
import CheckboxGroup from './checkboxgroup';
import RadioGroup from './radiogroup';
import DatePicker from './datepicker';
import RangePicker from './rangepicker';
import TimePicker from './timepicker';

// list
import Editable from './editable';

import { registerFieldPlugin } from './plugin';
import { Switch, Checkbox, Input, InputNumber, Rate, Radio, Slider } from 'antd';

function registerFieldPlugins() {
  registerFieldPlugin('input')(Input);
  registerFieldPlugin('textarea')(Input.TextArea);
  registerFieldPlugin('password')(Input.Password);
  registerFieldPlugin('inputnumber')(InputNumber);
  registerFieldPlugin('rate')(Rate);
  registerFieldPlugin('switch')(Switch);
  registerFieldPlugin('checkbox')(Checkbox);
  registerFieldPlugin('radio')(Radio);
  registerFieldPlugin('slider')(Slider);

  registerFieldPlugin('cascader')(Cascader, true);
  registerFieldPlugin('treeselect')(TreeSelect, true);
  registerFieldPlugin('datepicker')(DatePicker);
  registerFieldPlugin('rangepicker')(RangePicker);
  registerFieldPlugin('timepicker')(TimePicker);
  registerFieldPlugin('upload')(Upload, false, true);
  registerFieldPlugin('select')(Select, true);
  registerFieldPlugin('checkboxgroup')(CheckboxGroup, true);
  registerFieldPlugin('radiogroup')(RadioGroup, true);

  registerFieldPlugin('editable')(Editable, false, true);
}

export { Select, CheckboxGroup, TimePicker, DatePicker, RangePicker, Cascader, TreeSelect, RadioGroup, Upload, registerFieldPlugins, registerFieldPlugin };
