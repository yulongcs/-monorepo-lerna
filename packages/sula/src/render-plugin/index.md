---
title: 渲染插件
---

## 组件

```jsx
import React from 'react';
import { Card, Text, Icon, Button } from './';
import { IconPlugin } from './icon'
import { ButtonPlugin }  from './button';

export default () => {
  const [text, setText] = React.useState('文字');
  const btnRef = React.useRef(null);
  const iconRef = React.useRef(null);

  return (
    <div>
      <Card title="Head">Body</Card>
      <Text type="danger">hello</Text>
      <Icon type="tablet" theme="filled" />
      <Icon type="appstore" />
      <Icon type="appstore" onClick={() => {}} />
      <Icon type="car" onClick={() => {}} />
      <Icon type="coffee" onClick={() => {}} />
      <Icon type="appstore" disabled onClick={() => {}} />
      <Icon
        type="tablet"
        theme="filled"
        text={text}
        onClick={() => {
          setText(text ? undefined : '文字');
          console.log('点击');
        }}
      />
      <Button ref={btnRef} onClick={() => {
        btnRef.current.showLoading();
        setTimeout(() => {
          btnRef.current.hideLoading();
        }, 2000)
      }}>转2秒</Button><br />

      <IconPlugin
        text="转起来"
        config={{
          action: [
            ctx => {
              ctx.icon.showLoading();
              setTimeout(() => {
                ctx.icon.hideLoading();
              }, 2000);
            },
            () => {
              console.log('next action');
            },
            () => {
              console.log('finally action');
            },
          ]
        }}
        autoLoading={false}
        type="appstore"
      />
    </div>
  );
};
```

## 插件

```jsx
import React from 'react';
import sula from '../core';

export default () => {
  return (
    <div>
      {sula.render(
        'text',
        {},
        {
          props: {
            type: 'danger',
            children: 'danger',
          },
        },
      )}
      {
        sula.render(
        'icon',
        {},
        {
          props: {
            type: 'appstore',
          },
        },
      )
      }
      {
        sula.render(
        'tag',
        {},
        {
          props: {
            children: 'hello',
          },
        },
      )
      }
      {
        sula.render(
          'progress',
          {},
          {
            props: {
              percent: 30,
            },
          },
        )
      }
    </div>
  );
};
```

<code title="button" src="./demos/button.tsx" />
