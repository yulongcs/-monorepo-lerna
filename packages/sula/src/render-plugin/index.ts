import Card from './card';
import Text from './text';
// import Icon from './icon';
import Icon, { IconPlugin } from './icon';
import { Tag, Badge, Progress, Input } from 'antd';
import Button, { ButtonPlugin, LinkPlugin } from './button';

// only plugin
import RowSelection from './rowSelection';
import { registerRenderPlugin } from './plugin';

function registerRenderPlugins() {
  registerRenderPlugin('card')(Card);
  registerRenderPlugin('div')('div');
  registerRenderPlugin('inputgroup')(Input.Group);
  registerRenderPlugin('text')(Text);
  registerRenderPlugin('button', ['autoLoading'])(ButtonPlugin, true);
  registerRenderPlugin('link', ['autoLoading'])(LinkPlugin, true);
  registerRenderPlugin('icon', ['autoLoading'])(IconPlugin, true);

  registerRenderPlugin('rowSelection')(RowSelection, true);
  registerRenderPlugin('tag')(Tag);
  registerRenderPlugin('badge')(Badge);
  registerRenderPlugin('progress')(Progress);
}

export { Card, Text, Icon, Button, registerRenderPlugins, registerRenderPlugin };
