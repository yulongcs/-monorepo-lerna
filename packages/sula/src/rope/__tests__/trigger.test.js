/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';
import {
  getLazyCtx,
  triggerPlugin,
  triggerFieldPlugin,
  toArrayActions,
  triggerRenderPlugin,
  normalizeConfig,
} from '../triggerPlugin';
import { registerRenderPlugins } from '../../render-plugin';
import { registerFieldPlugins } from '../../field-plugin';

describe('sula trigger plugin', () => {
  describe('getLayyCtx', () => {
    it('no params', () => {
      expect(getLazyCtx()).toBeNull();
      const fn = jest.fn(() => ({ form: 'f', table: 't' }));
      expect(
        getLazyCtx({
          ctxGetter: fn,
          other: 'o',
        }),
      ).toEqual({ form: 'f', other: 'o', table: 't' });
      expect(fn).toHaveBeenCalled();
      expect(getLazyCtx({ ctxGetter: 'a', other: 'o' })).toEqual({ other: 'o' });
    });
  });

  describe('toArrayActions', () => {
    it('basic', () => {
      const actionsChain = [];
      const actions = [
        {
          type: 'first',
          finish: {
            type: 'second',
            finish: {
              type: 'third',
            },
          },
        },
      ];

      toArrayActions(actions, actionsChain);
      expect(actionsChain).toEqual([{ type: 'first' }, { type: 'second' }, { type: 'third' }]);

      const actionsChain2 = [];
      const actions2 = [
        [
          {
            type: 'a-1',
          },
          [
            {
              type: 'a-2-1',
            },
            {
              type: 'a-2-2',
            },
            {
              type: 'a-2-3',
              finish: {
                type: 'a-2-3-1',
              },
            },
          ],
        ],
        {
          type: 'b-1',
          finish: {
            type: 'b-1-1',
          },
        },
      ];
      toArrayActions(actions2, actionsChain2);
      expect(actionsChain2).toEqual([
        { type: 'a-1' },
        { type: 'a-2-1' },
        { type: 'a-2-2' },
        { type: 'a-2-3' },
        { type: 'a-2-3-1' },
        { type: 'b-1' },
        { type: 'b-1-1' },
      ]);
    });
  });

  describe('triggerPlugin', () => {
    registerRenderPlugins();
    registerFieldPlugins();

    it('function type', () => {
      expect(triggerPlugin('field', {}, { type: () => <div>abc</div> })).toEqual(<div>abc</div>);
    });

    it('basic object', () => {
      expect(
        triggerPlugin('render', {}, { type: 'text', props: { children: 'abc' } }),
      ).toMatchSnapshot();
    });

    it('template string', () => {
      expect(
        triggerPlugin(
          'field',
          { a: 'input!', disabled: true },
          {
            type: 'input',
            props: {
              placeholder: 'please #{a}',
            },
            funcProps: {
              disabled: (ctx) => ctx.disabled,
            },
          },
          null,
          true,
        ),
      ).toMatchSnapshot();
    });

    it('skip function props', () => {
      const fn = jest.fn();
      expect(
        triggerPlugin(
          'field',
          { holder: 'input!', disabled: true },
          {
            type: 'input',
            props: {
              placeholder: 'please #{holder}',
              onClick: fn,
            },
          },
          {
            skipFuncObjKeys: ['props'],
          },
          true,
        ),
      ).toMatchSnapshot();
      expect(fn).not.toHaveBeenCalled();
    });
  });

  describe('triggerFieldPlugin', () => {
    it('baisc', () => {
      const fn = jest.fn();
      expect(
        triggerFieldPlugin(
          { holder: 'input!' },
          {
            type: 'input',
            props: {
              placeholder: 'plese #{holder}',
              onClick: fn,
            },
          },
        ),
      ).toMatchSnapshot();
      expect(fn).not.toHaveBeenCalled();
    });
  });

  describe('normalizeConfig', () => {
    it('string && function', () => {
      expect(normalizeConfig('abc')).toEqual({ type: 'abc' });
      expect(
        normalizeConfig(() => {
          return 'abc';
        }).type(),
      ).toEqual('abc');
    });

    it('has type', () => {
      expect(normalizeConfig({ props: {} }, 'abc')).toEqual({
        type: 'abc',
        props: {},
      });
    });

    it('basic config', () => {
      expect(normalizeConfig({ type: 'abc', props: {} })).toEqual({
        type: 'abc',
        props: {},
      });
      expect(normalizeConfig(<div>abc</div>)).toEqual(<div>abc</div>);
    });

    describe('triggerRenderPlugin', () => {
      it('null', () => {
        expect(triggerRenderPlugin({}, [])).toBeNull();
      });

      it('not have action', () => {
        const text = triggerRenderPlugin(
          {},
          {
            type: 'text',
            props: {
              children: 'hello world',
              className: 't1',
            },
          },
        );
        expect(text).toMatchSnapshot();
      });

      it('basic', () => {
        const fn = jest.fn();
        const text = triggerRenderPlugin(
          {},
          {
            type: 'text',
            props: {
              children: 'hello world',
              className: 't1',
            },
            action: fn,
          },
        );
        const wrapper = mount(text);
        expect(fn).not.toHaveBeenCalled();
        expect(text).toMatchSnapshot();
        wrapper.find('.t1').at(0).simulate('click');
        expect(fn).toHaveBeenCalled();
      });

      it('only one child render', () => {
        const fn = jest.fn();
        const text = triggerRenderPlugin({}, [
          {
            type: 'text',
            props: {
              children: 'hello',
              className: 't1',
            },
            action: fn,
          },
        ]);
        const wrapper = mount(text);
        expect(fn).not.toHaveBeenCalled();
        expect(text).toMatchSnapshot();
        wrapper.find('.t1').at(0).simulate('click');
        expect(fn).toHaveBeenCalled();
      });

      it('multiple child render', () => {
        const fn1 = jest.fn();
        const fn2 = jest.fn();
        const text = triggerRenderPlugin({}, [
          {
            type: 'text',
            props: {
              children: 'hello',
              className: 't1',
            },
            action: fn1,
          },
          {
            type: 'text',
            props: {
              children: 'world',
              className: 't2',
            },
            action: fn2,
          },
        ]);
        const wrapper = mount(text);
        expect(fn1).not.toHaveBeenCalled();
        expect(fn2).not.toHaveBeenCalled();
        expect(text).toMatchSnapshot();
        wrapper.find('.t1').at(0).simulate('click');
        expect(fn1).toHaveBeenCalled();
        wrapper.find('.t2').at(0).simulate('click');
        expect(fn2).toHaveBeenCalled();
      });
    });
  });
});
