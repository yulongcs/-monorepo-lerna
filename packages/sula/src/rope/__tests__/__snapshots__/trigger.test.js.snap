// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sula trigger plugin normalizeConfig triggerRenderPlugin basic 1`] = `
<RopeContainer
  trigger={[Function]}
>
  <Text
    className="t1"
  >
    hello world
  </Text>
</RopeContainer>
`;

exports[`sula trigger plugin normalizeConfig triggerRenderPlugin multiple child render 1`] = `
<Space>
  <RopeContainer
    trigger={[Function]}
  >
    <Text
      className="t1"
    >
      hello
    </Text>
  </RopeContainer>
  <RopeContainer
    trigger={[Function]}
  >
    <Text
      className="t2"
    >
      world
    </Text>
  </RopeContainer>
</Space>
`;

exports[`sula trigger plugin normalizeConfig triggerRenderPlugin not have action 1`] = `
<Text
  className="t1"
>
  hello world
</Text>
`;

exports[`sula trigger plugin normalizeConfig triggerRenderPlugin only one child render 1`] = `
<RopeContainer
  trigger={[Function]}
>
  <Text
    className="t1"
  >
    hello
  </Text>
</RopeContainer>
`;

exports[`sula trigger plugin triggerFieldPlugin baisc 1`] = `
<Input
  onClick={[MockFunction]}
  placeholder="plese input!"
  type="text"
/>
`;

exports[`sula trigger plugin triggerPlugin basic object 1`] = `
<Text>
  abc
</Text>
`;

exports[`sula trigger plugin triggerPlugin skip function props 1`] = `
<Input
  disabled={true}
  onClick={[MockFunction]}
  placeholder="please input!"
  type="text"
/>
`;

exports[`sula trigger plugin triggerPlugin template string 1`] = `
<Input
  disabled={true}
  placeholder="please input!"
  type="text"
/>
`;
